langchain/vectorstores/matching_engine.py

```diff
@@ -0,0 +1,465 @@
+
+        Args:
+            texts: Iterable of strings to add to the vectorstore.
+            metadatas: Optional list of metadatas associated with the texts.
+            kwargs: vectorstore specific parameters.
+
+        Returns:
+            List of ids from adding the texts into the vectorstore.
+        """
+        logger.debug(f"Embedding documents.")
+        embeddings = self.embedding.embed_documents(list(texts))
+        jsons = []
+        ids = []
+        # Could be improved with async.
+        for embedding, text in zip(embeddings, texts):
+            id = str(uuid.uuid4())
+            ids.append(id)
+            jsons.append({
+                "id": id,
+                "embedding": embedding
+            })
+            self._upload_to_gcs(text, f"documents/{id}")
+
+        logger.debug(f"Uploaded {len(ids)} documents to GCS.")
+
+        # Creating json lines from the embedded documents.
+        result_str = "\n".join([json.dumps(x) for x in jsons])
+
+        filename_prefix = f"indexes/{uuid.uuid4()}"
+        filename = f"{filename_prefix}/{time.time()}.json"
+        self._upload_to_gcs(result_str, filename)
+        logger.debug(f"Uploaded updated json with embeddings to "
+                     f"{self.gcs_bucket_uri}/{filename}.")
+
+        self.index = self.index.update_embeddings(
+            contents_delta_uri=f"gs://{self.gcs_bucket_uri}/{filename_prefix}/"
+        )
+
+        logger.debug(f"Updated index with new configuration.")
+
+        return ids
+    
+    def _upload_to_gcs(self, data: str, gcs_location: str) -> None:
+        """Uploads data to gcs_location.
+
+        Args:
+            data: The data that will be stored.
+            gcs_location: The location where the data will be stored.
+        """
+        client = self._get_gcs_client()
+        bucket = client.get_bucket(self.gcs_bucket_uri)
+        blob = bucket.blob(gcs_location)
+        blob.upload_from_string(data)
+
+    def _get_gcs_client(self) -> "storage.Client":
+        """Lazily creates a GCS client.
+
+        Returns:
+            A configured GCS client.
+        """
+
+        from google.cloud import storage
+
+        if self.gcs_client is None:            
+            self.gcs_client = storage.Client(
+                credentials=self.credentials, 
+                project=self.project_id
+            )
+
+        return self.gcs_client
+
+    def similarity_search(
+        self, 
+        query: str, 
+        k: int = 4, 
+        **kwargs: Any
+    ) -> List[Document]:
+        """Return docs most similar to query.
+
+        Args:
+            query: The string that will be used to search for similar documents.
+            k: The amount of neighbors that will be retrieved.
+
+        Returns:
+            A list of k matching documents.
+        """
+
+        logger.debug(f"Embedding query {query}.")
+        embedding_query = self.embedding.embed_documents([query])
+
+        index_id = None
```

Comment:
It would be cleaner to extract lines 321 to 329 as a separate method `_get_index_id()`

Reply:
I like the suggestion